src/common/tests_dmg_helpers.c

```diff
@@ -94,6 +95,136 @@ cmd_string(const char *cmd_base, char *args[], int argcount)
 	return cmd_str;
 }
 
+static void
+log_stderr_pipe(int fd)
+{
+	char	buf[512];
+	char	*full_msg = NULL;
+	ssize_t	len = 0;
+
+	D_DEBUG(DB_TEST, "reading from stderr pipe\n");
+	while (1) {
+		ssize_t n;
+		ssize_t old_len = len;
+		char	*tmp;
+
+		n = read(fd, buf, sizeof(buf));
+		if (n == 0)
+			break;
+		if (n < 0) {
+			D_ERROR("read from stderr pipe failed: %s\n", strerror(errno));
+			break;
+		}
+
+		len = len + n;
+		D_REALLOC(tmp, full_msg, old_len, len);
+		if (tmp == NULL) {
+			D_ERROR("reading from stderr pipe: can't realloc tmp with size %ld\n",
+				len);
+			break;
+		}
+
+		full_msg = tmp;
+		strncpy(&full_msg[old_len], buf, n);
+	}
+
+
+	D_DEBUG(DB_TEST, "done reading stderr pipe\n");
+	close(fd);
+
+	if (full_msg == NULL) {
+		D_INFO("no stderr output\n");
+		return;
+	}
+
+	D_DEBUG(DB_TEST, "stderr: %s\n", full_msg);
+	D_FREE(full_msg);
+}
+
+static int
+run_cmd(const char *command, int *outputfd)
+{
+	int rc = 0;
+	int child_rc = 0;
+	int child_pid;
+	int stdoutfd[2];
+	int stderrfd[2];
+
+	D_DEBUG(DB_TEST, "dmg cmd: %s\n", command);
+
+	/* Create pipes */
+	if (pipe(stdoutfd) == -1) {
+		rc = daos_errno2der(errno);
+		D_ERROR("failed to create stdout pipe: %s\n", strerror(errno));
+		return rc;
+	}
+
+	if (pipe(stderrfd) == -1) {
+		rc = daos_errno2der(errno);
+		D_ERROR("failed to create stderr pipe: %s\n", strerror(errno));
+		close(stdoutfd[0]);
+		close(stdoutfd[1]);
+		return rc;
+	}
+
+	child_pid = fork();
+	if (child_pid == -1) {
+		rc = daos_errno2der(errno);
+		D_ERROR("failed to fork: %s\n", strerror(errno));
+		return rc;
+	}
+
+	if (child_pid == 0) {
+		/* child doesn't need the read end of the pipes */
+		close(stdoutfd[0]);
+		close(stderrfd[0]);
+
+		D_DEBUG(DB_TEST, "running dmg command\n");
+
+		if (dup2(stdoutfd[1], STDOUT_FILENO) == -1) {
+			D_ERROR("failed to dup stdout pipe: %s\n", strerror(errno));
+			_exit(daos_errno2der(errno));
+		}
+
+		if (dup2(stderrfd[1], STDERR_FILENO) == -1) {
+			D_ERROR("failed to dup stderr pipe: %s\n", strerror(errno));
+			_exit(daos_errno2der(errno));
+		}
+
+		close(stdoutfd[1]);
+		close(stderrfd[1]);
+
+		if (system(command) == -1) {
```

Comment:
why system rather than exec* ?

Reply:
It was simpler than breaking up the string into substrings to generate the argv array for execvp. Since this is only used by test code, I figured refactoring around reformatting the inputs wasn't that important.